Pytest ipdoctest plugin

Kojoley opened the pull request:
The plugin is based on Pytest doctest.py.

I am very confident that no tests are missing in Pytest discovery (except one bogus and two phantom duplicates, the iptest vs pytest difference can be found here https://gist.github.com/Kojoley/00ccf0aa22097a06cf2bb8d0c7ff78e8/revisions).

The doctest `doctest_who` was changed to do a soft reset, because a hard reset somehow causes failures in other tests that do `run_cell(x, store_history=True); assert _ip.user_ns['_'] == x`.

Kojoley committed:
Pytest ipdoctest plugin

Based on Pytest doctest.py with changes ported from Nose ipdoctest plugin.

Kojoley committed:
Fix xfail polyfill signature

Kojoley committed:
CI: Colorize and widen Pytest summary

Pytest shows some error messages only in summary while also truncating lines there.

Carreau commented on IPython/testing/plugin/pytest_ipdoctest.py:
```diff
@@ -0,0 +1,789 @@
+class MultipleDoctestFailures(Exception):
+    def __init__(self, failures: Sequence["doctest.DocTestFailure"]) -> None:
+        super().__init__()
+        self.failures = failures
```
note to self, reviewed until roughly here.

Kojoley replied:
I can split the commit into two commits: copying a Pytest's doctest module and the modifications to make it ipdoctest-capable. That should make it easier to review and to sync with Pytest in the future if needed.

Kojoley replied:
I went ahead and done that. I could also split off doctest->ipdoctest renaming and darker pleasing changes as separate commits if it will be helpful.

Carreau replied:
Thanks, it's ok, and we can put new files in darker ignore file if that helps.

Kojoley committed:
Pytest ipdoctest plugin base

This is a copy of Pytest doctest.py without any changes.

Kojoley commented:
I have identified the reason of strange failures when I don't disable those three doctests, it is because `exec` call adds its result into `buildins` module under `_` key. Why on earth it does that? The documentation keeps silence too. The iptest is unaffected by that because Nose doctest plugin wipes out `_` from `builtins` (`delattr(builtin_mod, self._result_var)`) while Pytest doctest does not.

```python
import builtins
assert '_' not in dir(builtins), 'before'
exec(compile('4\n', 'asd.py', "single", 0, True), {'__builtins__': {}})
assert '_' not in dir(builtins), 'after'
```
```
4
Traceback (most recent call last):
  File "...", line 4, in <module>
    assert '_' not in dir(builtins), 'after'
AssertionError: after
```

Kojoley commented:
I think the last test failure is uncovered longstanding bug, hard reset in doctests seems to screw things up in some way, and it was not observed in `iptest` because other regular tests happens to make hard reset too in the needed moment.

`iptest` runs tests in this order:
```
IPython/core/tests/test_logger.py::test_logstart_unicode ... ok
IPython/core/tests/test_magic.py::test_reset_out ... ok
IPython/core/tests/test_magic.py::IPython.core.tests.test_magic.doctest_who ... ok
```
If I force pytest to run them in that order -- everything is ok.
```
pytest -vv IPython/core/tests/test_logger.py::test_logstart_unicode IPython/core/tests/test_magic.py::test_reset_out IPython/core/tests/test_magic.py::IPython.core.tests.test_magic.doctest_who
```

I have workarounded the bug by making a soft reset instead in the doctest.

Kojoley committed:
Add builtins._ doctest

Defining builtins._ should not break anything outside the doctest while also should be working as expected inside the doctest.

Carreau commented:
I'm not surprised there are test ordering issues, feel free to make them as "skip" and "to fix later".

> it is because `exec` call adds its result into `builtins` module under `_` key

There is some behavior that try to keep the last result in `_` for when users forgot to assign a expensive result. It should be done via a shadow namespace now, and I guess the assigning to builtins has not been removed.

Kojoley commented:
> I'm not surprised there are test ordering issues, feel free to make them as "skip" and "to fix later".
> 
> > it is because `exec` call adds its result into `builtins` module under `_` key
> 
> There is some behavior that try to keep the last result in `_` for when users forgot to assign a expensive result. It should be done via a shadow namespace now, and I guess the assigning to builtins has not been removed.

Nah, it is just CPython `doctest` temporary restores the original display hook `sys.displayhook = sys.__displayhook__` and then calls `exec(compile(..., "single", ...), ...)` which calls `sys.displayhook` with the result (`sys.__displayhook__(result)` is essentially `builtin._ = result; print(result, file=sys.stdout)`).
